mbuild/recipes/monolayer.py

```diff
@@ -41,8 +42,30 @@ def __init__(self, surface, chain, backfill=None, pattern=None, tile_x=1,
         if pattern is None:  # Fill the surface.
             pattern = mb.Random2DPattern(len(tiled_compound.referenced_ports()))
 
-        # Attach chains to specified binding sites. Remaining sites get a backfill.
-        chains, backfills = pattern.apply_to_compound(guest=chain,
-                host=self['tiled_surface'], backfill=backfill, **kwargs)
-        self.add(chains)
+        if isinstance(chains, mb.Compound):
+            chains = [chains]
+
+        n_chains = len(pattern.points)
+
+        # Attach chains of each type to binding sites based on specified fractions
+        for i,chain in enumerate(chains[:-1]):
+
+            # Create sub-pattern for this chain type
+            subpattern = deepcopy(pattern)
+            n_points = round(fractions[i] * n_chains)
+            points = subpattern.points[np.random.choice(subpattern.points.shape[0], n_points, replace=False)]
+            subpattern.points = points
+
+            # Remove now-occupied points from overall pattern
+            pattern.points = np.array([point for point in pattern.points.tolist() if point not in subpattern.points.tolist()])
+
+            # Attach chains to the surface
+            attached_chains = subpattern.apply_to_compound(guest=chain,
+                              host=self['tiled_surface'], backfill=None, **kwargs)
+            self.add(attached_chains)
+
+        # Attach final chain type. Remaining sites get a backfill.
```

Comment:
I think you can get this all in the main loop and not replicate this. You just need a check to see if you're on the last one and then pass `backfill` instead of `None`.